.emacs.d/config.org
#+BEGIN_SRC emacs-lisp :comments no :tangle yes
  ;; -*- lexical-binding: t; -*-
#+END_SRC

#+TITLE: Heartmacs
#+PROPERTY: header-args :results silent :tangle yes :comments both :noweb yes
#+PROPERTY: header-args:emacs-lisp+ :lexical t
#+TAGS: keybinding | { os : macos | linux } | GTD
#+todo: TODO(t@) | DONE(d!)


* Toggle debug
:PROPERTIES:
:header-args: :tangle no
:END:

When having issues, set this to ~:tangle yes~
#+BEGIN_SRC emacs-lisp :tangle no
  (setq debug-on-error t)
#+END_SRC

If you need to figure out when a variable is modified, try using the
following snippet:
#+BEGIN_SRC emacs-lisp :tangle no
  (debug-on-variable-change 'evil-want-Y-yank-to-eol)
#+END_SRC

* Suppress warnings

Because some warnings give a lot of annoying pop-ups on startup, we
can disable showing them immediately. They will still show up in the
~*Warnings*~ buffer, they will just not show automatically.

#+BEGIN_SRC emacs-lisp
  (use-package emacs
    :custom
    (warning-suppress-types '((comp) (use-package) (:warning))))
#+END_SRC

* Utils for config

** Execute macro on kvps

A few times I've had to execute a macro on a list of key-value pairs
(usually to create some sort of bindings). To make this more
manageable, let's create a macro for it.

#+BEGIN_SRC emacs-lisp
  (defmacro heartman/execute-macro-on-kvp-list (macro &rest kvps)
    "Execute macro on the given list of args.

  MACRO is any macro. KVPS is a list of key-value pairs to use, formatted as a flat list of alternating keys and values (e.g. `(key1 value1 key2 value2)'). The MACRO is called once for with each pair as argument."
    (let (forms)
      (while kvps
        (let ((key (pop kvps)))
          (when (not kvps)
            (error "Couldn't execute macro. The provided args are unbalanced."))
          (let ((value (pop kvps)))
            (push `(,macro ,key ,value) forms))))
      `(progn ,@(nreverse forms))))
#+END_SRC

** Run a hook function once, then remove it

Taken from this [[https://emacs.stackexchange.com/questions/3323/is-there-any-way-to-run-a-hook-function-only-once][StackExchange question thread about running a hook function once]].

#+BEGIN_SRC emacs-lisp
  (defmacro add-hook-run-once (hook function &optional append local)
    "Like add-hook, but remove the hook after it is called"
    (let ((sym (gensym "#once")))
      `(progn
         (defun ,sym ()
           (remove-hook ,hook ',sym ,local)
           (funcall ,function))
         (add-hook ,hook ',sym ,append ,local))))
#+END_SRC

*** Run only on first graphical frame

Based on the run hook function once macro, we can ensure that certain things are only run on the first graphical frame. This is necessary because when emacs runs in daemon mode, some things won't work as expected. For instance, it does not have access to X fonts.

#+BEGIN_SRC emacs-lisp
  (defmacro heartman/on-gui-start (f)
    "Run F on first graphical frame."
    `(if (daemonp)
         (add-hook-run-once
          'server-after-make-frame-hook
          ,f)
       (funcall ,f)))
#+END_SRC



* Package setup

** Package archives

I'll sometimes want to use local packages. I'll put them in a
directory local to my emacs config:
#+BEGIN_SRC emacs-lisp
  (add-to-list 'load-path
               (expand-file-name
                "elisp"
                'user-emacs-directory))
#+END_SRC

* General
:PROPERTIES:
:REPO:     https://github.com/noctuid/general.el
:END:

General simplifies key binding.
#+BEGIN_SRC emacs-lisp
  (use-package general
    :config
    (general-auto-unbind-keys))
#+END_SRC

** Prefixes
:LOGBOOK:
- State "TODO"       from "TODO"       [2020-12-15 Tue 15:02] \\
  Rewrite this using macros to avoid repetitions.
- State "TODO"       from              [2020-08-14 Fri 21:25]
:END:

Similarly to how spacemacs has the leader key and a number of
mnemonic bindings, it'd be nice to do something similar.

I'd define a standard prefix, possibly a function key, and then spin
off from there. I'd like separate ones for
- things :: various apps and utilities. Elfeed, notmuch, org-agenda,
  notes, etc
- lsp / prog :: a key that gets you access to all bindings specified
  for the major prog mode you're working in.
- errors :: flycheck and or flyspell. Probably a hydra that lets you
  go from error to error (or just suggestions), correct
  them, or view all in a separate window


Let's create a couple prefixes:
#+BEGIN_SRC emacs-lisp
  (defconst my-format-buffer "<f1>")

  (defmacro create-definer (name key)
    "Create a constant and an associated definer.

  Use NAME to generate the name of the constant and definer, and
  use KEY for the definer prefix."
    (let ((constant-name (intern (format "my-%s-prefix" name)))
          (def-name (intern (format "my-%s-def" name))))
      `(progn
         (defconst ,constant-name ,key)
         (general-create-definer ,def-name :prefix ,constant-name))))

  (heartman/execute-macro-on-kvp-list
   create-definer
   "link" "C-x l"
   "flycheck" "C-c e"
   "lsp" "<f6>"
   "app" "<f5>"
   "package" "C-x p"
   "file" "C-x f")
#+END_SRC

* OS/system variables

#+BEGIN_SRC emacs-lisp
  (defconst on-macos (eq system-type 'darwin) "t if the current system is a darwin system (running macOS)")
  (defconst on-linux (eq system-type 'gnu/linux) "t if the current system is a linux system")
  (defconst use-exwm (and (display-graphic-p) on-linux) "t if exwm should be activated")
  (defconst native-comp (require 'comp nil t) "non-nil if Emacs uses native compilation")
#+END_SRC

* macOS binding

#+BEGIN_SRC emacs-lisp
  (when on-macos
    (setq mac-option-modifier 'super
          mac-command-modifier 'meta)
    (general-def
      "s-<backspace>" 'backward-kill-word
      "s-<delete>" 'kill-word
      "s-<kp-delete>" 'kill-word))
#+END_SRC

* Emacs

** Visual

These changes relate to how Emacs itself looks, including frames,
windows, and the like.

*** Frames

Get rid of tool bars and menu bars:
#+BEGIN_SRC emacs-lisp
  (menu-bar-mode -1)
  (tool-bar-mode -1)
#+END_SRC

*** Windows

Let's get rid of scroll bars:
#+BEGIN_SRC emacs-lisp
  (set-scroll-bar-mode nil)
#+END_SRC

I don't care much for fringes, so let's decrease the size a bit.

#+BEGIN_SRC emacs-lisp
  (set-fringe-mode nil)
#+END_SRC


*** Theme

To load a theme in Emacs, the default way is to use `M-x
load-theme`. This works well except for the fact that you need to
confirm that you really want to load it unless it's been added to
the 'safe themes' custom list. Also, Emacs doesn't disable the
modifications from a previous theme when you load a new one, so if
the previous theme defines more faces than the new one, some of
these will remain. To get around that we'll disable the old theme
if it is not the same as the new one.

#+BEGIN_SRC emacs-lisp
  (defun heartman/load-theme (theme)
    "Load custom theme named THEME from its file and enable it.

  Calls out to Emacs' own `load-theme' under the hood."
    (interactive
     (list
      (intern
       (completing-read "Load custom theme: "
                        (mapcar #'symbol-name
                                (custom-available-themes))))))

    (let ((old-theme (car custom-enabled-themes)))
      (when (not (eq theme old-theme))
        (message "Disabling theme %s" old-theme theme)
        (disable-theme old-theme))
      (load-theme theme t)
      ;; (when (string-match "^doom-" (prin1-to-string theme))
      ;;   ;; add an extra box around the mode line to make it easier to
      ;;   ;; distinguish between active and inactive windows.
      ;;   (let ((modeline-box-width 2))
      ;;     (custom-set-faces
      ;;      `(mode-line ((t (:box
      ;;                       (:color ,(doom-color 'magenta)
      ;;                               :line-width ,modeline-box-width)))))
      ;;      `(mode-line-inactive ((t (:box
      ;;                                (:color ,(doom-color 'modeline-bg)
      ;;                                        :line-width ,modeline-box-width)))))
      ;;      '(header-line ((t (:box nil))))
      ;;      `(child-frame-border ((t (:background ,(doom-color 'fg-alt))))))))
      (message "Loaded theme %s" theme)))
#+END_SRC


**** one-themes
:PROPERTIES:
:REPO:     https://github.com/balajisivaraman/emacs-one-themes
:header-args: :tangle no
:END:

I've been stuck with the Atom One Dark theme since I first saw it
a couple of years ago. This theme is the best Emacs-version I've
seen of it thus far.

However, the theme also affects certain font faces by default,
enabling mixed-pitch-mode for certain headings and changing other
font sizes. While it's not terrible, Emacs does better without
changing these things, so let's disable those changes.

#+BEGIN_SRC emacs-lisp
  (use-package one-themes
    :after org
    :init
    (setq emacs-one-use-variable-pitch nil
          emacs-one-scale-org-headlines nil
          emacs-one-height-minus-1 1.0
          emacs-one-height-plus-1 1.0
          emacs-one-height-plus-2 1.0
          emacs-one-height-plus-3 1.0
          emacs-one-height-plus-4 1.0))

#+END_SRC

**** doom themes
:PROPERTIES:
:REPO:     https://github.com/hlissner/emacs-doom-themes
:END:

#+BEGIN_SRC emacs-lisp
  (use-package doom-themes
    :config
    (setq doom-themes-enable-bold t
          doom-themes-enable-italic t)
    (doom-themes-visual-bell-config)
    (doom-themes-org-config))
#+END_SRC

**** Modus themes

Modus themes are another set of themes that I quite like. These are
high-contrast themes and come bundled with emacs.

#+BEGIN_SRC emacs-lisp
  (use-package emacs
    :custom
    (modus-themes-slanted-constructs t)
    (modus-themes-region 'bg-only)
    (modus-themes-bold-constructs t)
    (modus-themes-hl-line 'accented-background)
    (modus-themes-mode-line 'borderless-moody)
    (modus-themes-org-blocks 'gray-background)
    (modus-themes-scale-headings t)
    (modus-themes-variable-pitch-headings t))
#+END_SRC

**** Loading a theme

To try and stick to loading only one theme, let's do it here.

#+BEGIN_SRC emacs-lisp
  (heartman/load-theme 'doom-one)
#+END_SRC


*** Font

Next, let's set a list of my preferred fonts (in descending order
of preference), and check whether they exist on the system. If
they do, we'll set the first font we find as the default frame
font.

#+BEGIN_SRC emacs-lisp
  (require 'cl)

  (defvar heartman/default-font-height (if on-macos 200 100)
    "The height to use for the default face.")

  (defvar preferred-fonts '("JetBrains Mono"
                            "Cascadia Code"
                            "Victor Mono")
    "The list of preferred fonts in descending order. Emacs will
      use the first font it finds in this list.")

  (defun heartman/get-preferred-font ()
    (find-if
     (lambda (font) (member font (font-family-list))) preferred-fonts))

  (defun heartman/set-font ()
    (set-face-attribute 'default nil
                        :height heartman/default-font-height
                        :family (heartman/get-preferred-font)))

  (heartman/on-gui-start 'heartman/set-font)
#+END_SRC

**** Emoji

Because most fonts don't deal with Emoji, we have to set special
fonts for the emoji range. This snippet from [[http://ergoemacs.org/emacs/emacs_list_and_set_font.html][Ergoemacs' article
and fonts]] allows us to do just that.

When it's active, these emoji should render as expected:

😄 😱 😸 👸 👽 🙋 🥳 ❤️ ☺️ ✅ ⭕ ❌

Additionally, some emoji use [[https://emojipedia.org/emoji-zwj-sequence/][ZWJ sequences]]. At the moment, I have not
been able to render these correctly in Emacs, but that may have more
to do with the font than with Emacs. The emoji in the table below use
zero-width joiners. When they all display as single emoji, we'll know
be closer to full support.

| Emoji |
|-------|
| 👨‍👨‍👦    |
| 👨🏾‍🦱    |
| 🙋🏾‍♀️    |


#+BEGIN_SRC emacs-lisp
  ;; set font for emoji
  (set-fontset-font
   t
   '(#x1f300 . #x1fad0)
   (cond
    ((member "Noto Color Emoji" (font-family-list)) "Noto Color Emoji")
    ((member "JoyPixels" (font-family-list)) "JoyPixels")
    ((member "Noto Emoji" (font-family-list)) "Noto Emoji")
    ((member "Segoe UI Emoji" (font-family-list)) "Segoe UI Emoji")
    ((member "Apple Color Emoji" (font-family-list)) "Apple Color Emoji")
    ((member "Symbola" (font-family-list)) "Symbola")))
#+END_SRC

**** Mode line font size

We can make the mode line font be a little smaller than default to
give ourselves a little bit of extra space (based on [[https://emacs.stackexchange.com/a/1062][this answer on stack exchange]]):

#+BEGIN_SRC emacs-lisp
  (defvar heartman/mode-line-font-size .8
    "The size of the mode line font relative to the default.")

  (defun heartman/set-mode-line-font-size ()
    (interactive)
    (let ((faces '(mode-line
                   mode-line-buffer-id
                   mode-line-emphasis
                   mode-line-highlight
                   mode-line-inactive)))
      (mapc
       (lambda (face)
         (set-face-attribute face nil
                             :height heartman/mode-line-font-size))
       faces)))
#+END_SRC


*** Transparency

To set transparency, use:

#+BEGIN_SRC emacs-lisp
  (defun heartman/set-frame-opacity ()
    "Set the opacity of the current frame to OPACITY.

  OPACITY should be a value between 0 and 100 inclusive."
    (interactive)
    (let ((opacity (string-to-number (read-string "Opacity (0—100):"))))
      (message "opacity is %s" opacity)
      (set-frame-parameter (selected-frame) 'alpha `(,opacity ,opacity))))
#+END_SRC

#+BEGIN_SRC emacs-lisp :tangle no
  (defvar heartman/frame-opacity 93
    "The default frame opacity.")

  (add-to-list 'default-frame-alist `(alpha .  ,heartman/frame-opacity))
#+END_SRC

** Misc

*** ido mode

Inspired by the [[https://www.masteringemacs.org/article/introduction-to-ido-mode][Introduction to ido mode article]] from Mastering
Emacs, let's enable ido mode everywhere and also enable
flex-matching.
#+begin_src emacs-lisp
  (setq ido-enable-flex-matching t)
#+end_src

We can also use ido-mode for find file at point.
#+BEGIN_SRC emacs-lisp
  (setq ido-use-filename-at-point 'guess)
#+END_SRC

When entering a buffer name that doesn't exist when changing
buffers, let's let ido-mode create a new buffer.
#+BEGIN_SRC emacs-lisp
  (setq ido-create-new-buffer 'always)
#+END_SRC

*** Prompts

Taking another cue from Mastering Emacs ([[https://www.masteringemacs.org/article/disabling-prompts-emacs][Disabling Prompts in
Emacs]]), let's reduce the amount of prompts we come across in
Emacs:
#+BEGIN_SRC emacs-lisp
  (fset 'yes-or-no-p 'y-or-n-p)
  (setq confirm-nonexistent-file-or-buffer nil)
#+END_SRC

We can also disable prompts about killing buffers with live
processes attached.
#+BEGIN_SRC emacs-lisp
  (setq kill-buffer-query-funcions
        (remq 'process-kill-buffer-query-function
              kill-buffer-query-functions))
#+END_SRC

*** subword mode

Subword mode lets you navigate and act on parts of words when they
consist of multiple words put together, separated by capital
letters (as is common in programming), such as "GtkWindow" ("Gtk"
and "Window").

#+BEGIN_SRC emacs-lisp
  (dolist (mode '(prog-mode-hook text-mode-hook))
    (add-hook mode (lambda () (subword-mode 1))))
#+END_SRC

*** Hungry-backspace

Let's set up global hungry backspace keybindings too.

#+BEGIN_SRC emacs-lisp
  (general-def
    "C-c <DEL>" 'c-hungry-delete-backwards
    "C-c C-<DEL>" 'c-hungry-delete-backwards
    "C-c C-<delete>" 'c-hungry-delete-forward
    "C-c <deletechar>" 'c-hungry-delete-forward)
#+END_SRC


*** Startup

Let's get rid of splash screens and startup messages that we don't
need (also from [[https://www.masteringemacs.org/article/disabling-prompts-emacs][Disabling Prompts in Emacs]]):
#+BEGIN_SRC emacs-lisp
  (setq inhibit-startup-message t
        inhibit-startup-echo-area-message t)
#+END_SRC


*** recent files (recentf)                                       :keybinding:

Mickey Petersen suggests using the recent files package in his
[[https://www.masteringemacs.org/article/find-files-faster-recent-files-package][Find Files Faster with the Recent Files Package]], so we'll copy his
suggested configuration here.

#+BEGIN_SRC emacs-lisp
  (require 'recentf)

  (global-set-key (kbd "C-x C-r") 'ido-recentf-open)

  (recentf-mode t)

  (setq recentf-max-saved-items 50)

  (defun ido-recentf-open ()
    "Use `ido-completing-read to \\[find-file] a recent file"
    (interactive)
    (if (find-file (ido-completing-read "Find recent file: " recentf-list))
        (message "Opening file ...")
      (message "Aborting")))
#+END_SRC

*** Quality of life things

Additionally, let's enable a few things that just makes life
better in general:

Electric pair mode is great for automatic insertion of matching
braces. However, it clashes with smartparens, so we'll leave it
out for now.
#+BEGIN_SRC emacs-lisp
  ;; (electric-pair-mode)
#+END_SRC

Further, let's enable (relative) line numbers everywhere.
#+BEGIN_SRC emacs-lisp
  (setq display-line-numbers-type 'relative)
  (global-display-line-numbers-mode)
#+END_SRC

Make lines wrap visually if they're longer than what can fit in
the window. By default, visual line mode overrides a number of
functions that work on lines, but I find this confusing, so we'll
remove the key overrides it provides.
#+BEGIN_SRC emacs-lisp
  (global-visual-line-mode)
  (general-unbind
    'visual-line-mode-map
    [remap kill-line]
    [remap move-beginning-of-line]
    [remap move-end-of-line])
#+END_SRC

I prefer ending sentences with single spaces (even if I see the
benefits of using two, as recommended by Emacs), so let's also
tell Emacs that it's okay.
#+BEGIN_SRC emacs-lisp
  (setq sentence-end-double-space nil)
#+END_SRC

I also want files to auto-update if they change on disk, but I don't
want Emacs to notify me.
#+begin_src emacs-lisp
  (global-auto-revert-mode)

.emacs.d_new/config.el
;; Package archives
;;    Set up archives for packages:

;; [[file:~/.emacs.d_new/config.org::*Package archives][Package archives:1]]
(setq package-archives '(("gnu" . "https://elpa.gnu.org./packages/")
			 ("melpa" . "https://melpa.org/packages/")))
;; Package archives:1 ends here

;; use-package
;;    Before we do anything else, let's make sure we've got [[https://github.com/jwiegley/use-package][use-package]]
;;    ready to go! This bit of setup is a modified version of what Greg
;;    Stein's example from Caches to Caches post [[http://cachestocaches.com/2015/8/getting-started-use-package/][Getting started with
;;    use-package]].

;; [[file:~/.emacs.d_new/config.org::*use-package][use-package:1]]
(require 'package)
(setq package-enable-at-startup nil)
(package-initialize)

(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))

(eval-when-compile
  (require 'use-package)
  (setq use-package-always-ensure t))
;; use-package:1 ends here

;; Frames
;;     Get rid of tool bars and menu bars:

;; [[file:~/.emacs.d_new/config.org::*Frames][Frames:1]]
(menu-bar-mode -1)
(tool-bar-mode -1)
;; Frames:1 ends here

;; Windows
;;     Let's get rid of scroll bars:

;; [[file:~/.emacs.d_new/config.org::*Windows][Windows:1]]
(set-scroll-bar-mode nil)
;; Windows:1 ends here

;; Theme

;; [[file:~/.emacs.d_new/config.org::*Theme][Theme:1]]
(use-package one-themes
	     :config
	     (load-theme 'one-dark t))
;; Theme:1 ends here

;; Font
;;     The default font size is a bit small on my monitors, so we'll
;;     increase it sufficiently.

;; [[file:~/.emacs.d_new/config.org::*Font][Font:1]]
(set-face-attribute 'default nil :height 200)
;; Font:1 ends here

;; ido mode
;;     Inspired by the [[https://www.masteringemacs.org/article/introduction-to-ido-mode][Introduction to ido mode article]] from Mastering
;;     Emacs, let's enable ido mode everywhere and also enable
;;     flex-matching.

;; [[file:~/.emacs.d_new/config.org::*ido mode][ido mode:1]]
(setq ido-enable-flex-matching t
      ido-everywhere t)
(ido-mode 1)
;; ido mode:1 ends here



;; We can also use ido-mode for find file at point.

;; [[file:~/.emacs.d_new/config.org::*ido mode][ido mode:2]]
(setq ido-use-filename-at-point 'guess)
;; ido mode:2 ends here



;; When entering a buffer name that doesn't exist when changing
;; buffers, let's let ido-mode create a new buffer.

;; [[file:~/.emacs.d_new/config.org::*ido mode][ido mode:3]]
(setq ido-create-new-buffer 'always)
;; ido mode:3 ends here

;; Prompts
;;     Taking another cue from Mastering Emacs ([[https://www.masteringemacs.org/article/disabling-prompts-emacs][Disabling Prompts in
;;     Emacs]]), let's reduce the amount of prompts we come across in
;;     Emacs:

;; [[file:~/.emacs.d_new/config.org::*Prompts][Prompts:1]]
(fset 'yes-or-no-p 'y-or-n-p)
(setq confirm-nonexistent-file-or-buffer nil)
;; Prompts:1 ends here



;; We can also disable prompts about killing buffers with live
;; processes attached.

;; [[file:~/.emacs.d_new/config.org::*Prompts][Prompts:2]]
(setq kill-buffer-query-funcions
      (remq 'process-kill-buffer-query-function
	    kill-buffer-query-functions))
;; Prompts:2 ends here

;; Startup
;;     Let's get rid of splash screens and startup messages that we don't
;;     need (also from [[https://www.masteringemacs.org/article/disabling-prompts-emacs][Disabling Prompts in Emacs]]):

;; [[file:~/.emacs.d_new/config.org::*Startup][Startup:1]]
(setq inhibit-startup-message t
      inhibit-startup-echo-area-message t)
;; Startup:1 ends here

;; recent files (recentf)                                       :keybinding:
;;     Mickey Petersen suggests using the recent files package in his
;;     [[https://www.masteringemacs.org/article/find-files-faster-recent-files-package][Find Files Faster with the Recent Files Package]], so we'll copy his
;;     suggested configuration here.


;; [[file:~/.emacs.d_new/config.org::*recent files (recentf)][recent files (recentf):1]]
(require 'recentf)

(global-set-key (kbd "C-x C-r") 'ido-recentf-open)

(recentf-mode t)

(setq recentf-max-saved-items 50)

(defun ido-recentf-open ()
  "Use `ido-completing-read to \\[find-file] a recent file"
  (interactive)
  (if (find-file (ido-completing-read "Find recent file: " recentf-list))
      (message "Opening file ...")
    (message "Aborting")))
;; recent files (recentf):1 ends here

;; Quality of life things
;;     Additionally, let's enable a few things that just makes life
;;     better in general:

;; [[file:~/.emacs.d_new/config.org::*Quality of life things][Quality of life things:1]]
(electric-pair-mode)
;; Quality of life things:1 ends here



;; Further, let's enable line numbers everywhere and make them relative.

;; [[file:~/.emacs.d_new/config.org::*Quality of life things][Quality of life things:2]]
(global-display-line-numbers-mode)
(setq-default display-line-numbers 'relative)
;; Quality of life things:2 ends here



;; Make lines wrap visually if they're longer than what can fit in
;; the window.

;; [[file:~/.emacs.d_new/config.org::*Quality of life things][Quality of life things:3]]
(global-visual-line-mode)
;; Quality of life things:3 ends here



;; I prefer ending sentences with single spaces (even if I see the
;; benefits of using two, as recommended by Emacs), so let's also
;; tell Emacs that it's okay.

;; [[file:~/.emacs.d_new/config.org::*Quality of life things][Quality of life things:4]]
(setq sentence-end-double-space nil)
;; Quality of life things:4 ends here



;; I also want files to auto-update if they change on disk.

;; [[file:~/.emacs.d_new/config.org::*Quality of life things][Quality of life things:5]]
(global-auto-revert-mode)
;; Quality of life things:5 ends here



;; Trailing whitespace is something I prefer to avoid, so let's
;; delete that on save.

;; [[file:~/.emacs.d_new/config.org::*Quality of life things][Quality of life things:6]]
(add-hook 'before-save-hook 'delete-trailing-whitespace)
;; Quality of life things:6 ends here



;; For text modes, let's always use auto-fill mode and orgtbl-mode:

;; [[file:~/.emacs.d_new/config.org::*Quality of life things][Quality of life things:7]]
(defun my-text-mode-hook ()
  (auto-fill-mode)
  (orgtbl-mode))
(add-hook 'text-mode-hook 'my-text-mode-hook)
;; Quality of life things:7 ends here

;; Removing suspend-frame                                      :keybinding:
;;      To disable suspend-frame (I don't think I've ever meant to
;;      activate it), we'll instead replace it by a message saying it has
;;      been unmapped.

;; [[file:~/.emacs.d_new/config.org::*Removing suspend-frame][Removing suspend-frame:1]]
(global-set-key [remap suspend-frame]
  (lambda ()
    (interactive)
    (message "This keybinding is disabled (was 'suspend-frame')")))
;; Removing suspend-frame:1 ends here

;; Don't pause redisplay on input events
;;      According to [[https://www.masteringemacs.org/article/improving-performance-emacs-display-engine][Improving the performance of Emacs's Display Engine?]]
;;      by Mickey Petersen, Emacs defaults to pausing all display
;;      redrawing on any input. This may have been useful previously, but
;;      is not necessary anymore.

;; [[file:~/.emacs.d_new/config.org::*Don't pause redisplay on input events][Don't pause redisplay on input events:1]]
(setq redisplay-dont-pause t)
;; Don't pause redisplay on input events:1 ends here

;; Garbage collection
;;     Inspired by the [[https://github.com/lewang/flx][flx's readme]], we can increase the GC threshold
;;     significantly from the default. In addition to when using flx,
;;     this might also come in handy in other situations.

;; [[file:~/.emacs.d_new/config.org::*Garbage collection][Garbage collection:1]]
(setq gc-cons-threshold 20000000)
;; Garbage collection:1 ends here

;; macOS                                                            :macos:
;;      On macOS, I want the CMD key to act as Meta, and the Alt key as super, because this works better with their location on the keyboard and their corresponding keys on Linux.

;; [[file:~/.emacs.d_new/config.org::*macOS][macOS:1]]
(when (eq system-type 'darwin)
  (setq mac-option-modifier nil
        mac-command-modifier 'meta))
;; macOS:1 ends here

;; Ivy                                                            :keybinding:
;;   Following the [[https://oremacs.com/swiper/][Ivy Documentation]], let's install Ivy (and Counsel and
;;   Swiper) using counsel. We'll also replace normal isearch with Ivy's
;;   ~swiper-isearch~ and the regular ~yank-pop~ functionality with
;;   ~counsel-yank-pop~.

;;   For searches, I also prefer using a fuzzier regex pattern than the
;;   default, so this is set up by assigning ~ivy--regex-fuzzy~.


;; [[file:~/.emacs.d_new/config.org::*Ivy][Ivy:1]]
(use-package counsel
  :demand
  :config
  (ivy-mode 1)
  (setq ivy-use-virtual-buffers t
	ivy-count-format "(%d/%d)"
	ivy-re-builders-alist '((swiper-isearch . ivy--regex-plus)
				(t . ivy--regex-fuzzy))
	ivy-wrap t)
  :bind (("C-S-s" . swiper-isearch)
	 ("C-M-y" . counsel-yank-pop)))
;; Ivy:1 ends here

;; Flycheck

;; [[file:~/.emacs.d_new/config.org::*Flycheck][Flycheck:1]]
(use-package flycheck
  :ensure t
  :config (global-flycheck-mode))
;; Flycheck:1 ends here

;; TODO Company                                                   :keybinding:
;;   Completion is important! [[https://github.com/company-mode/company-mode][Company-mode repo]]

;;   We need to set this up properly. In addition to wanting to use
;;   tab-n-go, I also want to be able to use TAB to expand completions
;;   with function arguments (like Rust Analyzer).

;;   In addition to the basic company-mode configuration, I also use
;;   [[https://github.com/company-mode/company-mode/blob/master/company-tng.el][company-tng]] to get YCMD-like behavior.  This requires a little extra
;;   bit of configuration to get set up.


;; [[file:~/.emacs.d_new/config.org::*Company][Company:1]]
(defun my-company-tng-setup ()
  (setq company-require-match nil
	company-frontends '(company-tng-frontend
			    company-pseudo-tooltip-frontend
			    company-echo-metadata-frontend))
  (let ((keymap company-active-map))
    (define-key keymap [return] nil)
    (define-key keymap (kbd "RET") nil)))
(use-package company
  :config
  (global-company-mode)
  (global-set-key (kbd "C-<f5>") 'company-complete)
  (my-company-tng-setup)
  (setq company-idle-delay 0.2
	company-selection-wrap-around t))
;; Company:1 ends here



;; When using varying font sizes within a buffer, this can make the
;; company completion dropdown misaligned (with itself, even). Using
;; the [[https://github.com/tumashu/company-posframe][company-posframe package]] appears to fix this by putting
;; completions in a separate frame.

;; [[file:~/.emacs.d_new/config.org::*Company][Company:2]]
(use-package company-posframe
  :after company
  :config (company-posframe-mode 1))
;; Company:2 ends here



;; To make completions a bit smoother, I use [[https://github.com/PythonNut/company-flx][company-flx]] to allow for
;; fuzzy matching when company uses the company-capf backend. I'll also
;; add ~fuzzy~ to the list of completion styles.

;; [[file:~/.emacs.d_new/config.org::*Company][Company:3]]
(use-package company-flx
  :after company
  :config
  (company-flx-mode 1)
  (add-to-list 'completion-styles 'fuzzy))
;; Company:3 ends here

;; TODO Magit                                                     :keybinding:
;;   Because it's simply one of the best git experiences out there, of
;;   course I want to enable and use Magit:

;; [[file:~/.emacs.d_new/config.org::*Magit][Magit:1]]
(use-package magit
  :bind (("C-x g" . magit-status)
	 ("C-x M-g" . magit-dispatch)))
;; Magit:1 ends here



;; Additionally, to smooth out the workflow with GitHub and GitLab,
;; let's also use Forge.

;; Note: this is currently commented out because Forge seems to throw
;; errors when I've already got spacemacs set up in
;; ~user-emacs-directory~. Comment this back in when I switch over.

;; [[file:~/.emacs.d_new/config.org::*Magit][Magit:2]]
;; (use-package forge
;;   :after magit
;;   :demand
;;   :config
;;   (add-to-list 'forge-alist
;; 	       '("gitlab.intility.no" "gitlab.intility.no/api/v4" "gitlab.intility.no" forge-gitlab-repository))
;;   :bind (:map magit-mode-map
;; 	      ("C-c M-w" . forge-copy-url-at-point-as-kill)))
;; Magit:2 ends here

;; Rainbow delimiters
;;   Rainbow delimiters make it much easier to read a lot of code, so
;;   let's set them up.

;; [[file:~/.emacs.d_new/config.org::*Rainbow delimiters][Rainbow delimiters:1]]
(use-package rainbow-delimiters
  :hook ((prog-mode text-mode) . rainbow-delimiters-mode))
;; Rainbow delimiters:1 ends here

;; Nix mode                                                       :keybinding:
;;   :PROPERTIES:
;;   :REPO:     [[https://github.com/NixOS/nix-mode/][GitHub]]
;;   :END:

;;   Using NixOS as my main OS and Nix shells for dev environments on
;;   other platforms, it's quite important to get this mode set up.

;;   Of note: nix-mode seems to have a dependency on json-mode, so make
;;   sure to only initialize after said mode.

;; [[file:~/.emacs.d_new/config.org::*Nix mode][Nix mode:1]]
(use-package nix-mode
  :after json-mode
  :mode "\\.nix\\'"
  :bind (:map nix-mode-map
	      ("C-c C-f" . nix-format-buffer)))
;; Nix mode:1 ends here

;; JSON mode
;;   :PROPERTIES:
;;   :REPO:      [[https://github.com/joshwnj/json-mode][GitHub]]
;;   :END:
;;   For JSON support and, more specifically, for Nix mode, which seems
;;   to require this.


;; [[file:~/.emacs.d_new/config.org::*JSON mode][JSON mode:1]]
(use-package json-mode)
;; JSON mode:1 ends here

;; TODO Change inner                                              :keybinding:
;;   Note: this could do with some improvements to also accept closing parens,
;;   braces, brackets, etc., and to allow certain shortcuts, such as ~b~ for
;;   ~parens~. Should actually be fairly doable. This has been reported previously
;;   ([[https://github.com/magnars/change-inner.el/issues/8][issue]]) and been deemed not worth doing. Another option is [[https://gist.github.com/alphapapa/fd7edf8104215028f3da][this gist]] by
;;   Alphapapa, which could work well.

;;   To emulate some of Vim's behavior, we'll use [[https://github.com/magnars/change-inner.el][change-inner.el]]:

;; [[file:~/.emacs.d_new/config.org::*Change inner][Change inner:1]]
(use-package change-inner
  :bind (("M-i" . change-inner)
	 ("M-o" . change-outer)))
;; Change inner:1 ends here

;; Spell checker
;;   Enable regular spell checking in all text modes and prog type spell
;;   checking in prog modes:

;; [[file:~/.emacs.d_new/config.org::*Spell checker][Spell checker:1]]
(add-hook 'prog-mode-hook 'flyspell-prog-mode)
(add-hook 'text-mode-hook 'flyspell-mode)
;; Spell checker:1 ends here

;; Markdown mode
;;   :PROPERTIES:
;;   :REPO:     [[https://github.com/jrblevin/markdown-mode][GitHub]]
;;   :END:


;; [[file:~/.emacs.d_new/config.org::*Markdown mode][Markdown mode:1]]
(use-package markdown-mode
  :mode (("\\.md\\'" . gfm-mode))
  :after
  (edit-indirect)
  :config
  (setq markdown-asymmetric-header t))
;; Markdown mode:1 ends here

;; Edit-indirect
;;   :PROPERTIES:
;;   :REPO:     [[https://github.com/Fanael/edit-indirect][GitHub]]
;;   :END:

;;   To edit arbitrary code in separate buffers (the same way org mode
;;   does), the package edit-indirect exist. This is required by markdown
;;   mode to edit source code blocks in separate buffers.


;; [[file:~/.emacs.d_new/config.org::*Edit-indirect][Edit-indirect:1]]
(use-package edit-indirect)
;; Edit-indirect:1 ends here
